MakieCore/src/conversion.jl

```diff
@@ -30,3 +30,103 @@ conversion_trait(::Type{<: Heatmap}) = DiscreteSurface()
+
+Converts the arguments to their correct type for the Plot type.
+Throws apropriate errors if it can't convert!
+"""
+function convert_arguments_typed end
+
+
+function convert_arguments_typed(P::Type{<: AbstractPlot}, @nospecialize(args...))
+    convert_arguments_typed(conversion_trait(P), args...)
+end
+
+function convert_arguments_typed(::CT, @nospecialize(args...)) where {CT <: ConversionTrait}
+    return convert_arguments_typed(CT, args...)
+end
+
+function convert_arguments_typed(ct::Type{<: ConversionTrait}, @nospecialize(args...))
+    # We currently just fall back to not doing anything if there isn't a convert_arguments_typed defined for certain plot types.
+    # This makes `@convert_target` an optional feature right now.
+    # It will require a bit more work to make it mandatory, and we have to think more about how to integrate it with user recipes
+    return args
+end
+
+convert_arguments_typed(::NoConversion, args...) = args
+
+"""
+    @convert_target(expr)
+
+Allows to define a conversion target for a plot type, so that convert_arguments can be checked properly, if it converts to the correct types.
+
+Usage:
+```Julia
+@convert_target struct PointBased{N} # Can be the Plot type or a conversion Trait
+    positions::AbstractVector{Point{N, Float32}}
+end
+```
+"""
+macro convert_target(struct_expr)
+    if !Meta.isexpr(struct_expr, :struct)
+        error("Expression must be `struct Target; fields...; end`")
+    else
+        target_name = struct_expr.args[2]
+
+        if Meta.isexpr(target_name, :curly)
+            target_name, targs... = target_name.args
+        else
+            targs = ()
+        end
+
+        body = struct_expr.args[3]
+        convert_expr = []
+        converted = Symbol[]
+        all_fields = filter(x-> !(x isa LineNumberNode), body.args)
+        if any(x-> !Meta.isexpr(x, :(::)), all_fields)
+            error("All fields need to be of type `name::Type`. Found: $(all_fields)")
+        end
+        names = map(x-> x.args[1], all_fields)
+        types = map(x-> :($(x.args[2]) where {$(targs...)}), all_fields)
+        for (name, TargetType) in zip(names, types)
+            conv_name = Symbol("converted_$name")
+            push!(converted, conv_name)
+            # We always add the where clause, since it's too complicated to match the static type parameters to the types that use them.
+            # This seems to work well, since Julia drops any unecessary where clause/ type parameter, while lowering.
+            # TODO figure out a way to drop the where close... Eval is the only thing I found, but shouldn't be used here.
+
+            expr = quote
+                # Unions etc don't work well with `convert(T, x)`, so we try not to convert if already the right type!
+                if $name isa $TargetType
+                    $conv_name = $name
+                else
+                    $conv_name = try
+                        convert($TargetType, $name)
```

Comment:
I want to come back here later and use e.g. `convert_attribute` here, or some of the newer versions we're using in `Block` or the new `Attribute` type.
For now, I decided to stick with `convert` for two reasons:
* the `convert` here is kind of optional, since `convert_arguments` should already convert to the right types
* it's mainly a softer version of just asserting the type, and therefore `convert` does the job well in throwing the appropriate errors even for union types